### PR TITLE
[llvm] Convert byref args to i64 stack type

### LINKED ISSUE #16808: LLVM failed for String.GetHashCode
`String.GetHashCode` calls `Marvin.ComputeHash32` which can not be converted to LLVM, a simple repro:

```csharp
using System;
using System.Runtime.CompilerServices;

unsafe class Program
{
    static void Main(string[] args)
    {
        byte[] a = new byte[100];
        ComputeHash32(ref a[0]);
    }
    
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static unsafe void ComputeHash32(ref byte data)
    {
        for (int i=0; i<1; i++)
            data = ref Unsafe.Add(ref data, 1);
            // or:
            // data = ref Unsafe.AddByteOffset(ref data, (IntPtr)1);
    }
}
```
`--llvm` output contains:

```
LLVM failed for 'Program.ComputeHash32': incoming phi arg type mismatch
```
It explains why `string.GetHashCode` is still slow even if LLVM mode (JIT) is enabled

### PR DESCRIPTION
**Pull request by filipnavara**
Fixes #16808

### CODE REVIEW

File: mono/mini/mini-llvm.c [hunk 1]
```diff
@@ -3579,7 +3579,10 @@ emit_entry_bb (EmitContext *ctx, LLVMBuilderRef builder)
 			ctx->addresses [reg] = LLVMGetParam (ctx->lmethod, pindex);
 			break;
 		default:
-			ctx->values [reg] = convert_full (ctx, ctx->values [reg], llvm_type_to_stack_type (cfg, type_to_llvm_type (ctx, ainfo->type)), type_is_unsigned (ctx, ainfo->type));
+			if (ainfo->type->byref)
+				ctx->values [reg] = convert_full (ctx, ctx->values [reg], IntPtrType(), FALSE);
+			else
+				ctx->values [reg] = convert_full (ctx, ctx->values [reg], llvm_type_to_stack_type (cfg, type_to_llvm_type (ctx, ainfo->type)), type_is_unsigned (ctx, ainfo->type));
```
Comment (filipnavara) [inline mono/mini/mini-llvm.c]: Match the logic here:

https://github.com/mono/mono/blob/9608debd7bdf94fe27258d7365d7c04d1c648d90/mono/mini/ssa.c#L440-L443

**Comment by filipnavara** [comment]
@monojenkins build OS X i386

**Comment by filipnavara** [comment]
@monojenkins build Windows x64 FullAOT+LLVM

**Comment by filipnavara** [comment]
@vargaz @imhameed Can you review please? I rewrote the original hack into something that I believe to be correct.

**Comment by vargaz** [comment]
Would prefer:
https://github.com/mono/mono/pull/16874

**Comment by kg** [comment]
I'm not super familiar with our LLVM infrastructure but do we pipe our wasm through llvm? If so, the correct stack type would be pointer not i64, since in wasm pointers are 32-bit

**Comment by filipnavara** [comment]
> Would prefer: #16874

That is exactly what I originally used. Only downside is that I was not sure if it properly covers all the cases...

**Comment by filipnavara** [comment]
>  If so, the correct stack type would be pointer not i64, since in wasm pointers are 32-bit

@kg it would actually use whatever the native pointer type is. I should have been more clear about that. The problem was that it used pointer type (i64*) where pointer-sized integer type was expected.

**Pull request closed without merging**